### DISCUSSION TITLE
Transform time series with missing data points?

### DISCUSSION THREAD

**Discussion by bertzruph** [vega/altair#3669] [Q&A]
Hi,

I have a time series where I want to plot the individual data points and a 5-day rolling average, e.g:

Date | Value
:-- | :--
25/10/2024 | 218.8
26/10/2024 | 218
27/10/2024 | 219
28/10/2024 | 220.2
29/10/2024 | 221.2
30/10/2024 | 217.8
31/10/2024 | 216.2
02/11/2024 | 220.2

Notice there is no value for Nov 1st. I'd like my average line to represent the average of the values falling within the five-day period instead of the previous five data points. I can do this in pandas using a resample method, but is there a way to achieve this using altair transformations?

Thanks

**Comment by joelostblom** [comment]
It sounds like [the imputation transform](https://altair-viz.github.io/user_guide/transform/impute.html) together with a window `frame` could be useful.

**Comment by bertzruph** [comment]
Hi Joel,

Thanks for responding. I've taken a look at the `transform_impute` examples in the docs and tried to adapt for my use, but there is still a gap in my chart. I think where my case differs is that there is no data at all for Nov 1st, whereas in the documentation examples at least one of the series is observed in each step. In this way Altair "knows" there is a space that needs to be filled. I think I will need to preprocess my table to insert a record for Nov 1st before passing to Altair.

Best wishes

**Reply by joelostblom** [reply]
Yes, I believe that is correct. I think you can also supply a numerical sequence to the `keyvals` parameter [as is shown in the Vega-Lite docs here](https://vega.github.io/vega-lite/docs/impute.html#specifying-the-key-values-to-be-imputed). I don't know if it supports timestamps directly, but you could always create a numerical index corresponding to your timestamp. You might also find it easier to impute for timestamps in the dataframe library you are using.
